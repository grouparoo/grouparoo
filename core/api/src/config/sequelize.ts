import { URL } from "url";
import { join } from "path";

// we want BIGINTs to be returned as JS integer types
require("pg").defaults.parseInt8 = true;

const databaseBaseName = "grouparoo";

export const DEFAULT = {
  sequelize: (config) => {
    let dialect = "postgres";
    let host = process.env.DB_HOST || "127.0.0.1";
    let port = process.env.DB_PORT || "5432";
    let database =
      process.env.DB_DATABASE ||
      `${databaseBaseName}_${config.process.env}${
        process.env.JEST_WORKER_ID ? "_" + process.env.JEST_WORKER_ID : ""
      }`;
    let username =
      process.env.DB_USER || process.env.CI ? "postgres" : undefined;
    let password = process.env.DB_PASS || undefined;

    // if your environment provides database information via a single JDBC-style URL
    // like mysql://username:password@hostname:port/default_schema
    const connectionURL =
      process.env.DATABASE_URL || process.env.MYSQL_URL || process.env.PG_URL;

    if (connectionURL) {
      const parsed = new URL(connectionURL);
      if (parsed.protocol) {
        dialect = parsed.protocol.slice(0, -1);
      }
      if (parsed.username) {
        username = parsed.username;
      }
      if (parsed.password) {
        password = parsed.password;
      }
      if (parsed.hostname) {
        host = parsed.hostname;
      }
      if (parsed.port) {
        port = parsed.port;
      }
      if (parsed.pathname) {
        database = parsed.pathname.substring(1);
      }
    }

    if (dialect === "postgresql") {
      dialect = "postgres";
    }

    return {
      autoMigrate: true,
      logging: false,
      dialect: dialect,
      port: parseInt(port),
      database: database,
      host: host,
      username: username,
      password: password,
      models: [join(__dirname, "..", "models")],
      migrations: [join(__dirname, "..", "migrations")],
      pool: {
        max: process.env.SEQUELIZE_POOL_SIZE
          ? parseInt(process.env.SEQUELIZE_POOL_SIZE)
          : 5,
        min: 0,
        acquire: 30000,
        idle: 10000,
      },
    };
  },
};

// for the sequelize CLI tool
module.exports.development = DEFAULT.sequelize({
  env: "development",
  process: {
    env: "development",
  },
});

module.exports.staging = DEFAULT.sequelize({
  env: "staging",
  process: {
    env: "staging",
  },
});

module.exports.production = DEFAULT.sequelize({
  env: "production",
  process: {
    env: "production",
  },
});
