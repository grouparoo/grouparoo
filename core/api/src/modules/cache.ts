import { api, cache } from "actionhero";
import { waitForLock } from "./locks";

export interface ObjectCacheMethod {
  (
    argument: {
      objectGuid: string;
      cacheKey: string;
      cacheDurationMs?: number;
      lock?: boolean; // send false to no lock
      read?: boolean; // send false to not use cache
      write?: boolean; // send false to not write to cache
      passthru?: boolean; // send true to not read or write or lock
    },
    methodToGetValue: { (): Promise<any> }
  ): Promise<any>;
}

export interface ObjectCacheInvalidateMethod {
  (argument: { objectGuid: string }): Promise<void>;
}

async function getCacheValue({
  valueKey,
  read,
}): Promise<{ cached: boolean; value: any }> {
  if (!read) {
    return { cached: false, value: undefined };
  }
  try {
    const cached = await cache.load(valueKey);
    return { cached: true, value: cached.value };
  } catch (error) {
    const message = (error?.message || "").toString();
    if (message === "Object not found" || message === "Object expired") {
      return { cached: false, value: undefined };
    }
    throw error;
  }
}

function makeObjectKey({ objectGuid }) {
  objectGuid = (objectGuid || "").toString().trim();
  if (!objectGuid) {
    throw new Error(`objectGuid required`);
  }

  // TODO: validate if real guid?

  return `objectcache:${objectGuid}`;
}

function makeBaseKey({ objectGuid, cacheKey }): string {
  const objectKey = makeObjectKey({ objectGuid });

  cacheKey = (cacheKey || "").toString().trim();
  if (!cacheKey) {
    throw new Error(`cacheKey required`);
  }

  return `${objectKey}:${cacheKey}`;
}

function useRedis() {
  const running = !!api?.redis?.clients;
  if (!running && process.env.NODE_ENV === "test") {
    return false;
  }
  return true;
}

export const objectCacheInvalidate: ObjectCacheInvalidateMethod = async ({
  objectGuid,
}) => {
  if (!useRedis()) {
    return;
  }

  const client = cache.client();
  const objectKey = makeObjectKey({ objectGuid });
  const prefix = `${cache.redisPrefix}${objectKey}`;
  const keys = await client.keys(prefix + "*");
  const jobs = [];
  keys.forEach((key: string) => {
    jobs.push(client.del(key));
  });
  await Promise.all(jobs);
};

export const objectCache: ObjectCacheMethod = async (
  {
    objectGuid,
    cacheKey,
    cacheDurationMs = 1000 * 10,
    lock = true,
    read = true,
    write = true,
    passthru = false,
  },
  methodToGetValue
) => {
  if (!useRedis()) {
    passthru = true;
  }

  if (passthru) {
    read = false;
    write = false;
    lock = false;
  }

  const baseKey = makeBaseKey({ objectGuid, cacheKey });
  const valueKey = `${baseKey}:value`;
  const lockKey = `${baseKey}:lock`;

  let { value, cached } = await getCacheValue({ valueKey, read });
  if (cached) {
    return value;
  }

  let releaseLock: Function = null;
  try {
    if (lock) {
      releaseLock = (await waitForLock(lockKey)).releaseLock;

      let { value, cached } = await getCacheValue({ valueKey, read });
      if (cached) {
        return value;
      }
    }

    const value = await methodToGetValue();
    if (write) {
      await cache.save(valueKey, value, cacheDurationMs);
    }
    return value;
  } finally {
    if (releaseLock) {
      await releaseLock();
    }
  }
};
