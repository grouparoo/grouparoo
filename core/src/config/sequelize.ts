import fs from "fs";
import { URL } from "url";
import { join, isAbsolute } from "path";
import { getParentPath, getPluginManifest } from "../modules/pluginDetails";
import { log } from "actionhero";
import cls from "cls-hooked";
import Sequelize from "sequelize";

const namespace = "sequelize";

declare module "actionhero" {
  export interface ActionheroConfigInterface {
    [namespace]: ReturnType<typeof DEFAULT[typeof namespace]>;
  }
}

// Opt Into CLS
// Learn more @ https://sequelize.org/master/manual/transactions.html and https://github.com/Jeff-Lewis/cls-hooked
// @ts-ignore
Sequelize.useCLS(cls.createNamespace("grouparoo-cls"));

// we want BIGINTs to be returned as JS integer types
require("pg").defaults.parseInt8 = true;

export const DEFAULT = {
  [namespace]: (config) => {
    const env = process.env.NODE_ENV ?? "development";
    let storage: string; //only for sqlite
    let dialect = process.env.DB_DIALECT;
    let host = process.env.DB_HOST;
    let port = process.env.DB_PORT;
    let database = process.env.DB_DATABASE;
    let username = process.env.DB_USER;
    let password = process.env.DB_PASS || undefined;
    let schema = process.env.DB_SCHEMA || undefined;
    let ssl: boolean | { [key: string]: any } = false;

    // if your environment provides database information via a single JDBC-style URL
    // like mysql://username:password@hostname:port/default_schema
    let parsed: { [key: string]: any } = {};
    const connectionURL =
      process.env.DATABASE_URL || process.env.MYSQL_URL || process.env.PG_URL;
    if (connectionURL) parsed = new URL(connectionURL);

    if (parsed.protocol) dialect = parsed.protocol.slice(0, -1).toLowerCase();
    if (dialect === "postgresql") dialect = "postgres";
    if (dialect === "psql") dialect = "postgres";

    /** POSTGRES */
    if (dialect === "postgres") {
      if (parsed.username) username = parsed.username;
      if (parsed.password) password = parsed.password;
      if (parsed.hostname) host = parsed.hostname;
      if (parsed.port) port = parsed.port;
      if (parsed.pathname) {
        database = parsed.pathname.substring(1);
        if (database.includes("/")) {
          const dbAndSchema = database.split("/");
          if (dbAndSchema[0].trim() !== "") database = dbAndSchema[0];
          if (dbAndSchema[1].trim() !== "") schema = dbAndSchema[1];
        }
      }
      const search_ssl = parsed.searchParams?.get("ssl");
      const search_sslmode = parsed.searchParams?.get("sslmode");
      if (search_ssl) ssl = search_ssl === "true";
      if (search_sslmode) {
        ssl = search_sslmode === "true" || search_sslmode === "required";
      }

      if (process.env.DATABASE_SSL?.toLowerCase() === "true") {
        ssl = true;
      }
      if (process.env.DATABASE_SSL_SELF_SIGNED?.toLowerCase() === "true") {
        ssl = { rejectUnauthorized: false };
      }
    }

    /** SQLITE */
    if (dialect === "sqlite") {
      storage = ":memory:";
      if (parsed.hostname || parsed.pathname) {
        storage = `${parsed.hostname}${parsed.pathname}`;
      }

      if (env === "test" && !parsed.hostname && !parsed.pathname) {
        storage = join(getParentPath(), `${database}.sqlite`);
      }

      if (storage === "memory") storage = ":memory:";

      // without a starting "/" we assume relative locations are against project root
      if (storage !== ":memory:" && !isAbsolute(storage)) {
        storage = join(getParentPath(), storage);
      }
    }

    /** Query Logging */
    function logging(message: string, time: number) {
      if (typeof message !== "string") return;
      log(message, "debug", { time });
    }

    /** Migration */
    function shouldAutoMigrate() {
      const runMode = process.env.GROUPAROO_RUN_MODE;
      const workers = parseInt(process.env.WORKERS || "0");

      if (["cli:apply"].includes(runMode)) {
        // doing it in the same transaction elsewhere
        return false;
      }
      if (["test", "development"].includes(env)) {
        // working locally, make sure to migrate
        return true;
      }
      if (["cli:config"].includes(runMode)) {
        // always migrate because don't have other workers
        return true;
      }
      if (workers > 0) {
        // in production, background servers do the migration
        return true;
      }
      return false;
    }

    /** Load plugin migrations */
    const { plugins } = getPluginManifest();
    const pluginMigrations = [];
    for (const plugin of plugins) {
      const path = join(plugin.path, "dist", "migrations");
      if (fs.existsSync(path)) {
        pluginMigrations.push(path);
      }
    }

    return {
      _toExpand: false,
      logging,
      benchmark: true,
      autoMigrate: shouldAutoMigrate(),
      dialect: dialect,
      port: parseInt(port),
      database: database,
      host: host,
      username: username,
      password: password,
      schema: schema,
      searchPath: schema,
      models: [join(__dirname, "..", "models")],
      migrations: [join(__dirname, "..", "migrations"), ...pluginMigrations],
      migrationLogLevel: storage === ":memory:" ? "debug" : "info",
      storage, // only used for sqlite
      dialectOptions: { ssl, prependSearchPath: schema !== undefined },
      transactionType: "DEFERRED",
      pool: {
        max:
          dialect === "sqlite"
            ? 1
            : Math.max(
                parseInt(process.env.SEQUELIZE_POOL_SIZE || "0"),
                parseInt(process.env.WORKERS || "0") + 1,
                1
              ),
        min: 1,
        acquire: 30000,
        idle: 10000,
        evict: 1000,
      },
      retry: {
        match: [
          Sequelize.ConnectionError,
          Sequelize.DatabaseError,
          Sequelize.TimeoutError,
          Sequelize.ConnectionTimedOutError,
        ],
        backoffBase: dialect === "sqlite" ? 1000 : 100,
        backoffExponent: dialect === "sqlite" ? 1.5 : 1.1,
        max: 5,
        timeout: undefined as number,
      },
    };
  },
};

// for the sequelize CLI tool
module.exports.development = DEFAULT.sequelize({
  env: "development",
  process: {
    env: "development",
  },
});

module.exports.staging = DEFAULT.sequelize({
  env: "staging",
  process: {
    env: "staging",
  },
});

module.exports.production = DEFAULT.sequelize({
  env: "production",
  process: {
    env: "production",
  },
});
