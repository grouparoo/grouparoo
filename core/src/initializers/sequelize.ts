import { Sequelize } from "sequelize-typescript";
import { api, config, Initializer, log } from "actionhero";
import { Migrations } from "../modules/migrations";
import { CLS } from "../modules/cls";

declare module "actionhero" {
  export interface Api {
    sequelize: Sequelize;
  }
}

export class SequelizeInitializer extends Initializer {
  constructor() {
    super();
    this.name = "sequelize";
    this.loadPriority = 201;
    this.stopPriority = 9000;
  }

  async initialize() {
    api.sequelize = new Sequelize(config.sequelize);
    await this.test();

    if (config.sequelize.autoMigrate) {
      await CLS.wrap(async () => {
        await Migrations.migrate(
          config.sequelize,
          api.sequelize,
          log,
          config.sequelize.migrationLogLevel
        );
      });
    }
  }

  async stop() {
    await api.sequelize.close();
  }

  async test() {
    let query = "SELECT NOW()";
    if (config.sequelize.dialect === "mssql") query = "SELECT GETDATE();";
    if (config.sequelize.dialect === "sqlite")
      query = "SELECT strftime('%s', 'now');";

    await api.sequelize.query(query);
  }
}
