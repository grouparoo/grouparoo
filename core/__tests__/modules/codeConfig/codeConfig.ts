import { helper } from "@grouparoo/spec-helper";
import {
  plugin,
  Property,
  App,
  Source,
  Schedule,
  Group,
  GroupRule,
  Destination,
  ApiKey,
  Team,
  TeamMember,
  Setting,
} from "../../../src";
import path from "path";
import { api, specHelper } from "actionhero";
import { Op } from "sequelize";
import { loadConfigDirectory } from "../../../src/modules/configLoaders";

describe("modules/codeConfig", () => {
  helper.grouparooTestServer({
    truncate: true,
    enableTestPlugin: true,
    resetSettings: true,
  });

  describe("initial config", () => {
    beforeAll(async () => {
      // manually run the initializer again after the server has started.
      // the test test-app plugin has been loaded
      api.codeConfig.allowLockedModelChanges = true;
      const { errors, seenIds, deletedIds } = await loadConfigDirectory(
        path.join(__dirname, "..", "..", "fixtures", "codeConfig", "initial")
      );
      expect(errors).toEqual([]);
      expect(seenIds).toEqual({
        apikey: ["website_key"],
        app: expect.arrayContaining(["data_warehouse", "events"]),
        destination: ["test_destination"],
        group: ["email_group"],
        property: expect.arrayContaining([
          "user_id",
          "email",
          "last_name",
          "first_name",
        ]),
        schedule: ["users_table_schedule"],
        source: ["users_table"],
        team: ["admin_team"],
        teammember: ["demo"],
      });
      expect(deletedIds).toEqual({
        apikey: [],
        app: [],
        destination: [],
        group: [],
        property: [],
        schedule: [],
        source: [],
        team: [],
        teammember: [],
      });
    });

    test("settings are updated", async () => {
      const setting = await plugin.readSetting("core", "cluster-name");
      expect(setting.value).toBe("Test Cluster");
      expect(setting.locked).toBe("config:code");
    });

    test("apps are created", async () => {
      const apps = await App.findAll({
        order: [["type", "asc"]],
      });
      expect(apps.length).toBe(2);

      expect(apps[0].id).toBe("events");
      expect(apps[0].name).toBe("Grouparoo Events");
      expect(apps[0].state).toBe("ready");
      expect(apps[0].locked).toBe("config:code");
      let options = await apps[0].getOptions();
      expect(options).toEqual({
        identifyingPropertyId: "user_id",
      });

      expect(apps[1].id).toBe("data_warehouse");
      expect(apps[1].name).toBe("Data Warehouse");
      expect(apps[1].state).toBe("ready");
      expect(apps[1].locked).toBe("config:code");
      options = await apps[1].getOptions();
      expect(options).toEqual({ fileId: "test-file-path.db" });
    });

    test("sources are created", async () => {
      const sources = await Source.findAll();
      expect(sources.length).toBe(1);
      expect(sources[0].id).toBe("users_table");
      expect(sources[0].appId).toBe("data_warehouse");
      expect(sources[0].name).toBe("Users Table");
      expect(sources[0].state).toBe("ready");
      expect(sources[0].locked).toBe("config:code");
      const options = await sources[0].getOptions();
      expect(options).toEqual({ table: "users" });
    });

    test("the bootstrapped property is created", async () => {
      const property = await Property.findOne({
        where: { directlyMapped: true },
      });
      expect(property.id).toBe("user_id");
      expect(property.key).toBe("userId");
      expect(property.type).toBe("integer");
      expect(property.unique).toBe(true);
      expect(property.identifying).toBe(true);
      expect(property.state).toBe("ready");
      expect(property.locked).toBe("config:code");
    });

    test("schedules are created", async () => {
      const schedules = await Schedule.findAll();
      expect(schedules.length).toBe(1);
      expect(schedules[0].id).toBe("users_table_schedule");
      expect(schedules[0].sourceId).toBe("users_table");
      expect(schedules[0].name).toBe("Users Table Schedule");
      expect(schedules[0].state).toBe("ready");
      expect(schedules[0].recurring).toBe(true);
      expect(schedules[0].recurringFrequency).toBe(900000);
      expect(schedules[0].locked).toBe("config:code");
    });

    test("properties are created", async () => {
      const rules = await Property.findAll();
      expect(rules.length).toBe(4);
      expect(rules.map((r) => r.key).sort()).toEqual([
        "email",
        "first name",
        "last name",
        "userId",
      ]);
      expect(rules.map((r) => r.sourceId).sort()).toEqual([
        "users_table",
        "users_table",
        "users_table",
        "users_table",
      ]);
      expect(rules.map((r) => r.state).sort()).toEqual([
        "ready",
        "ready",
        "ready",
        "ready",
      ]);
      expect(rules.map((r) => r.locked).sort()).toEqual([
        "config:code",
        "config:code",
        "config:code",
        "config:code",
      ]);
    });

    test("groups are created", async () => {
      const groups = await Group.findAll();
      expect(groups.length).toBe(1);
      expect(groups[0].id).toBe("email_group");
      expect(groups[0].name).toBe("People with Email Addresses");
      expect(groups[0].locked).toBe("config:code");
      const rules = await groups[0].getRules();
      expect(rules).toEqual([
        {
          key: "userId",
          match: "null",
          operation: { description: "is not equal to", op: "ne" },
          relativeMatchDirection: null,
          relativeMatchNumber: null,
          relativeMatchUnit: null,
          topLevel: false,
          type: "integer",
        },
        {
          key: "email",
          match: "%@%",
          operation: { description: "is like (case sensitive)", op: "like" },
          relativeMatchDirection: null,
          relativeMatchNumber: null,
          relativeMatchUnit: null,
          topLevel: false,
          type: "email",
        },
      ]);
    });

    test("destinations are created", async () => {
      const destinations = await Destination.findAll();
      expect(destinations.length).toBe(1);
      expect(destinations[0].id).toBe("test_destination");
      expect(destinations[0].appId).toBe("data_warehouse");
      expect(destinations[0].name).toBe("Test Destination");
      expect(destinations[0].syncMode).toBe("additive");
      expect(destinations[0].state).toBe("ready");
      expect(destinations[0].locked).toBe("config:code");
      const options = await destinations[0].getOptions();
      expect(options).toEqual({ table: "output" });
    });

    test("apiKeys are created", async () => {
      const apiKeys = await ApiKey.findAll();
      expect(apiKeys.length).toBe(1);
      expect(apiKeys[0].id).toBe("website_key");
      expect(apiKeys[0].name).toBe("web-api-key");
      expect(apiKeys[0].locked).toBe("config:code");
      expect(apiKeys[0].permissionAllRead).toBe(true);
      expect(apiKeys[0].permissionAllWrite).toBe(true);
    });

    test("teams are created", async () => {
      const teams = await Team.findAll();
      expect(teams.length).toBe(1);
      expect(teams[0].id).toBe("admin_team");
      expect(teams[0].name).toBe("Admin Team");
      expect(teams[0].locked).toBe("config:code");
      expect(teams[0].permissionAllRead).toBe(true);
      expect(teams[0].permissionAllWrite).toBe(true);
    });

    test("teamMembers are created", async () => {
      const teamMembers = await TeamMember.findAll();
      expect(teamMembers.length).toBe(1);
      expect(teamMembers[0].email).toEqual("demo@grouparoo.com");
      expect(teamMembers[0].firstName).toEqual("Example");
      expect(teamMembers[0].lastName).toEqual("Person");
      expect(await teamMembers[0].checkPassword("password")).toBe(true);
    });
  });

  describe("changed config", () => {
    beforeAll(async () => {
      api.codeConfig.allowLockedModelChanges = true;
      const { errors, seenIds, deletedIds } = await loadConfigDirectory(
        path.join(__dirname, "..", "..", "fixtures", "codeConfig", "changes")
      );
      expect(errors).toEqual([]);
      expect(seenIds).toEqual({
        apikey: ["website_key"],
        app: expect.arrayContaining(["data_warehouse", "events"]),
        destination: [],
        group: ["email_group"],
        property: expect.arrayContaining([
          "user_id",
          "last_name",
          "first_name",
          "email",
        ]),
        schedule: ["users_table_schedule"],
        source: ["users_table"],
        team: ["admin_team"],
        teammember: ["demo"],
      });
      expect(deletedIds).toEqual({
        apikey: [],
        app: [],
        destination: ["test_destination"],
        group: [],
        property: [],
        schedule: [],
        source: [],
        team: [],
        teammember: [],
      });
    });

    test("settings can be changed", async () => {
      const setting = await plugin.readSetting("core", "cluster-name");
      expect(setting.value).toBe("Test Cluster!!!");
    });

    test("changes to an app setting will be updated", async () => {
      const apps = await App.findAll({
        where: { type: { [Op.ne]: "events" } },
      });
      expect(apps.length).toBe(1);
      expect(apps[0].id).toBe("data_warehouse");
      expect(apps[0].name).toBe("Data Warehouse");
      expect(apps[0].state).toBe("ready");
      expect(apps[0].locked).toBe("config:code");
      const options = await apps[0].getOptions();
      expect(options).toEqual({ fileId: "new-file-path.db" });
    });

    test("property keys changes will be updated", async () => {
      const rules = await Property.findAll();
      expect(rules.length).toBe(4);
      expect(rules.map((r) => r.key).sort()).toEqual([
        "Email",
        "First Name",
        "Last Name",
        "userId",
      ]);
      expect(rules.map((r) => r.sourceId).sort()).toEqual([
        "users_table",
        "users_table",
        "users_table",
        "users_table",
      ]);
      expect(rules.map((r) => r.state).sort()).toEqual([
        "ready",
        "ready",
        "ready",
        "ready",
      ]);
      expect(rules.map((r) => r.locked).sort()).toEqual([
        "config:code",
        "config:code",
        "config:code",
        "config:code",
      ]);
    });

    test("groups can have changed names and rules", async () => {
      const groups = await Group.findAll();
      expect(groups.length).toBe(1);
      expect(groups[0].id).toBe("email_group");
      expect(groups[0].name).toBe("People who have Email Addresses");
      expect(groups[0].locked).toBe("config:code");
      const rules = await groups[0].getRules();
      expect(rules).toEqual([
        {
          key: "Email",
          match: "%@%",
          operation: { description: "is like (case sensitive)", op: "like" },
          relativeMatchDirection: null,
          relativeMatchNumber: null,
          relativeMatchUnit: null,
          topLevel: false,
          type: "email",
        },
      ]);
    });

    test("a removed destination will be deleted", async () => {
      const destinations = await Destination.scope(null).findAll();
      expect(destinations.length).toBe(1);
      const destination = destinations[0];
      expect(destination.state).toEqual("deleted");

      // we need to "wait" for the destination to be deleted to remove it's dependant models
      const foundTasks = await specHelper.findEnqueuedTasks(
        "destination:destroy"
      );
      await specHelper.runTask("destination:destroy", foundTasks[0].args[0]);
      await destination.reload();
      await destination.destroy();
    });

    test("changes to team permissions will be updated", async () => {
      const teams = await Team.findAll();
      expect(teams.length).toBe(1);
      expect(teams[0].id).toBe("admin_team");
      expect(teams[0].name).toBe("Admin Team (no write)");
      expect(teams[0].locked).toBe("config:code");
      expect(teams[0].permissionAllRead).toBe(true);
      expect(teams[0].permissionAllWrite).toBe(false);
    });

    test("a team member password can be changed", async () => {
      const teamMembers = await TeamMember.findAll();
      expect(teamMembers.length).toBe(1);
      expect(teamMembers[0].email).toEqual("demo@grouparoo.com");
      expect(teamMembers[0].firstName).toEqual("Example");
      expect(teamMembers[0].lastName).toEqual("Person");
      expect(await teamMembers[0].checkPassword("new-password")).toBe(true);
    });
  });

  describe("partially empty config", () => {
    beforeAll(async () => {
      api.codeConfig.allowLockedModelChanges = true;
      const { errors, seenIds, deletedIds } = await loadConfigDirectory(
        path.join(
          __dirname,
          "..",
          "..",
          "fixtures",
          "codeConfig",
          "partially-empty"
        )
      );
      expect(errors).toEqual([]);
      expect(seenIds).toEqual({
        apikey: [],
        app: ["data_warehouse"],
        destination: [],
        group: [],
        property: ["user_id", "email"],
        schedule: [],
        source: ["users_table"],
        team: [],
        teammember: [],
      });
      expect(deletedIds).toEqual({
        apikey: ["website_key"],
        app: ["events"],
        destination: [],
        group: ["email_group"],
        property: expect.arrayContaining(["last_name", "first_name"]),
        schedule: ["users_table_schedule"],
        source: [],
        team: ["admin_team"],
        teammember: ["demo"],
      });
    });

    afterAll(async () => {
      // pretending that the task had run to delete the group
      await Group.truncate();
      await GroupRule.truncate();
    });

    test("most objects will be deleted with an empty config file", async () => {
      expect(await App.count()).toBe(1);
      expect(await Source.count()).toBe(1);
      expect(await Schedule.count()).toBe(0);
      expect(await Destination.count()).toBe(0);
      expect(await Property.count()).toBe(2);
      expect(await ApiKey.count()).toBe(0);
      expect(await Team.count()).toBe(0);
      expect(await TeamMember.count()).toBe(0);
    });

    test("settings remain", async () => {
      expect(await Setting.count()).toBeGreaterThan(1);
    });

    test("groups will be moved to the deleted state and unlocked", async () => {
      const groups = await Group.scope(null).findAll();
      expect(groups.length).toBe(1);
      expect(groups[0].state).toBe("deleted");
      expect(groups[0].locked).toBe(null);
    });
  });

  describe("empty config", () => {
    beforeAll(async () => {
      api.codeConfig.allowLockedModelChanges = true;
      const { errors, seenIds, deletedIds } = await loadConfigDirectory(
        path.join(__dirname, "..", "..", "fixtures", "codeConfig", "empty")
      );
      expect(errors).toEqual([]);
      expect(seenIds).toEqual({
        apikey: [],
        app: [],
        destination: [],
        group: [],
        property: [],
        schedule: [],
        source: [],
        team: [],
        teammember: [],
      });
      expect(deletedIds).toEqual({
        apikey: [],
        app: ["data_warehouse"],
        destination: [],
        group: [],
        property: ["user_id", "email"],
        schedule: [],
        source: ["users_table"],
        team: [],
        teammember: [],
      });
    });

    test("all objects will be deleted with an empty config file", async () => {
      expect(await App.count()).toBe(0);
      expect(await Source.count()).toBe(0);
      expect(await Schedule.count()).toBe(0);
      expect(await Destination.count()).toBe(0);
      expect(await Group.count()).toBe(0);
      expect(await Property.count()).toBe(0);
      expect(await ApiKey.count()).toBe(0);
      expect(await Team.count()).toBe(0);
      expect(await TeamMember.count()).toBe(0);
    });

    test("settings remain", async () => {
      expect(await Setting.count()).toBeGreaterThan(1);
    });
  });

  describe("duplicate IDs", () => {
    test("config with duplicate IDs will not be applied", async () => {
      api.codeConfig.allowLockedModelChanges = true;
      const { errors } = await loadConfigDirectory(
        path.join(
          __dirname,
          "..",
          "..",
          "fixtures",
          "codeConfig",
          "duplicate-id"
        )
      );

      expect(errors[0]).toEqual("Duplicate ID values found: data_warehouse_a");
    });
  });

  describe("errors", () => {
    describe("plugin not installed", () => {
      beforeAll(async () => {
        api.codeConfig.allowLockedModelChanges = true;
      });

      test("missing plugin", async () => {
        const { errors } = await loadConfigDirectory(
          path.join(
            __dirname,
            "..",
            "..",
            "fixtures",
            "codeConfig",
            "error-not-installed"
          )
        );
        expect(errors[0]).toMatch(
          /Cannot find a \"missing-plugin\" connection available within the installed plugins. Current connections installed are:/
        );
      });
    });

    describe("bad ID", () => {
      beforeAll(async () => {
        api.codeConfig.allowLockedModelChanges = true;
      });

      test("space in ID", async () => {
        const { errors } = await loadConfigDirectory(
          path.join(__dirname, "..", "..", "fixtures", "codeConfig", "error-id")
        );
        expect(errors[0]).toMatch(/invalid id/);
      });
    });

    describe("app", () => {
      beforeAll(async () => {
        api.codeConfig.allowLockedModelChanges = true;
      });

      test("missing option", async () => {
        const { errors } = await loadConfigDirectory(
          path.join(
            __dirname,
            "..",
            "..",
            "fixtures",
            "codeConfig",
            "error-app"
          )
        );
        expect(errors[0]).toMatch(
          /fileId is required for a app of type test-plugin-app/
        );
      });
    });

    describe("source", () => {
      beforeAll(async () => {
        api.codeConfig.allowLockedModelChanges = true;
      });

      test("broken source", async () => {
        const { errors } = await loadConfigDirectory(
          path.join(
            __dirname,
            "..",
            "..",
            "fixtures",
            "codeConfig",
            "error-source"
          )
        );
        expect(errors[0]).toMatch(/Could not find object with ID: user_id/);
      });
    });

    describe("property", () => {
      beforeAll(async () => {
        api.codeConfig.allowLockedModelChanges = true;
      });

      test("broken property", async () => {
        const { errors } = await loadConfigDirectory(
          path.join(
            __dirname,
            "..",
            "..",
            "fixtures",
            "codeConfig",
            "error-property"
          )
        );
        expect(errors[0]).toMatch(
          /Could not find object with ID: missing_source/
        );
      });
    });

    describe("group", () => {
      beforeAll(async () => {
        api.codeConfig.allowLockedModelChanges = true;
      });

      test("broken group", async () => {
        const { errors } = await loadConfigDirectory(
          path.join(
            __dirname,
            "..",
            "..",
            "fixtures",
            "codeConfig",
            "error-group"
          )
        );
        expect(errors[0]).toMatch(
          /Could not find object with ID: missing_profile_property/
        );
      });
    });

    describe("broken team member", () => {
      beforeAll(async () => {
        api.codeConfig.allowLockedModelChanges = true;
      });

      test("errors will be thrown if the configuration is invalid", async () => {
        const { errors } = await loadConfigDirectory(
          path.join(
            __dirname,
            "..",
            "..",
            "fixtures",
            "codeConfig",
            "error-teamMember"
          )
        );
        expect(errors[0]).toMatch(/TeamMember.firstName cannot be null/);
      });
    });
  });

  describe("errors in a second run", () => {
    beforeEach(async () => {
      helper.truncate();
    });

    test("changing an app's type (missing) causes an error", async () => {
      // initial load
      await loadConfigDirectory(
        path.join(__dirname, "..", "..", "fixtures", "codeConfig", "initial")
      );

      // changes
      const { errors } = await loadConfigDirectory(
        path.join(
          __dirname,
          "..",
          "..",
          "fixtures",
          "codeConfig",
          "error-changed-type-missing"
        )
      );

      expect(errors.length).toBe(1);
      expect(errors[0]).toMatch(
        /Cannot find a \"foo\" app available within the installed plugins. Current apps installed are/
      );
    });

    test("changing an app's type (valid) causes an error", async () => {
      // initial load
      await loadConfigDirectory(
        path.join(__dirname, "..", "..", "fixtures", "codeConfig", "initial")
      );

      // changes
      const { errors } = await loadConfigDirectory(
        path.join(
          __dirname,
          "..",
          "..",
          "fixtures",
          "codeConfig",
          "error-changed-type-duplicate"
        )
      );

      expect(errors.length).toBe(1);
      expect(errors[0]).toMatch(/fileId is not an option for a manual app/);
    });
  });
});
