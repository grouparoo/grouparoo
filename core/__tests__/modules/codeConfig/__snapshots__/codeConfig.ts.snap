// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/codeConfig names should not be unqiue against deleted entries removed configs should be deleted before creating/updating 1`] = `
Object {
  "deletedIds": Object {
    "apikey": Array [
      "website_key",
    ],
    "app": Array [],
    "destination": Array [
      "test_destination",
    ],
    "group": Array [],
    "profile": Array [],
    "property": Array [
      "email",
      "first_name",
      "last_name",
    ],
    "schedule": Array [
      "users_table_schedule",
    ],
    "source": Array [],
    "team": Array [
      "admin_team",
    ],
    "teammember": Array [
      "demo",
    ],
  },
  "errors": Array [],
  "seenIds": Object {
    "apikey": Array [],
    "app": Array [
      "data_warehouse",
    ],
    "destination": Array [
      "data_warehouse_destination",
    ],
    "group": Array [
      "calculate_dates",
    ],
    "profile": Array [],
    "property": Array [
      "user_id",
    ],
    "schedule": Array [],
    "source": Array [
      "users_table",
    ],
    "team": Array [],
    "teammember": Array [],
  },
}
`;

exports[`modules/codeConfig names should not be unqiue against deleted entries removed configs should be deleted before creating/updating 2`] = `
Object {
  "deletedIds": Object {
    "apikey": Array [],
    "app": Array [
      "data_warehouse",
    ],
    "destination": Array [
      "data_warehouse_destination",
    ],
    "group": Array [
      "calculate_dates",
    ],
    "profile": Array [],
    "property": Array [
      "user_id",
    ],
    "schedule": Array [],
    "source": Array [
      "users_table",
    ],
    "team": Array [],
    "teammember": Array [],
  },
  "errors": Array [],
  "seenIds": Object {
    "apikey": Array [],
    "app": Array [
      "data_warehouse_2",
    ],
    "destination": Array [
      "data_warehouse_destination_2",
    ],
    "group": Array [
      "calculate_dates_2",
    ],
    "profile": Array [],
    "property": Array [
      "user_id_2",
    ],
    "schedule": Array [],
    "source": Array [
      "users_table_2",
    ],
    "team": Array [],
    "teammember": Array [],
  },
}
`;
